python/pyiceberg/io/s3fs.py

```diff
@@ -0,0 +1,235 @@
+        raise RuntimeError(f"Cannot retrieve object info: {self.location}")
+
+    def exists(self) -> bool:
+        """Checks whether the location exists"""
+        return self._s3.lexists(self.location)
+
+    def open(self) -> S3fsInputStream:
+        """Create an S3fsInputStream for reading the contents of the file
+
+        Returns:
+            S3fsInputStream: An S3fsInputStream instance for the file located at `self.location`
+        """
+        return S3fsInputStream(s3_object=self._s3.open(self.location, "rb"))
+
+
+class S3fsOutputFile(OutputFile):
+    """An output file implementation for the S3fsFileIO
+
+    Args:
+        location(str): An S3 URI
+
+    Attributes:
+        location(str): An S3 URI
+    """
+
+    def __init__(self, location: str, s3):
+        self._s3 = s3
+        super().__init__(location=location)
+
+    def __len__(self) -> int:
+        """Returns the total length of the file, in bytes"""
+        object_info = self._s3.info(self.location)
+        if object_info.get("Size"):  # s3fs versions seem inconsistent on the case used for size
+            return object_info["Size"]
+        elif object_info.get("size"):
+            return object_info["size"]
+        raise RuntimeError(f"Cannot retrieve object info: {self.location}")
+
+    def exists(self) -> bool:
+        """Checks whether the location exists"""
+        return self._s3.lexists(self.location)
+
+    def create(self, overwrite: bool = False) -> S3fsOutputStream:
+        """Create an S3fsOutputStream for reading the contents of the file
+
+        Args:
+            overwrite(bool): Whether to overwrite the file if it already exists
+
+        Returns:
+            S3fsOutputStream: A wrapper for writing to an s3 object located at self.location
+
+        Raises:
+            FileExistsError: If the file already exists at `self.location` and `overwrite` is False
+
+        Note:
+            If overwrite is set to False, a check is first performed to verify that the file does not exist.
+            This is not thread-safe and a possibility does exist that the file can be created by a concurrent
+            process after the existence check yet before the output stream is created. In such a case, the default
+            behavior will truncate the contents of the existing file when opening the output stream.
+        """
+        if not overwrite and self.exists():
+            raise FileExistsError(f"Cannot create file, file already exists: {self.location}")
+        self._s3.touch(self.location)
+        return S3fsOutputStream(s3_object=self._s3.open(self.location, "wb"))
+
+    def to_input_file(self) -> S3fsInputFile:
+        """Returns a new S3fsInputFile for the location at `self.location`"""
+        return S3fsInputFile(location=self.location, s3=self._s3)
+
+
+class S3fsFileIO(FileIO):
+    """A FileIO implementation that uses s3fs"""
+
+    def __init__(self, client_kwargs: Optional[dict] = None, **s3fs_filesystem_kwargs):
```

Comment:
I think we should have a convention for passing these arg groups. Catalogs accept an arbitrary property map and that is primarily what will be used to create `FileIO` instances. That means we should translate from a single map for all of these options.

For these two, I would create a `client` group that we pull from the properties dict and pass here, then pass the rest of the properties somewhere. We may want to also standardize the S3 args from Java and pass them correctly here. What do you think?